pkg/apps/packagekit.js

```diff
@@ -176,21 +191,47 @@ function refresh(origin_files, progress_cb) {
             update_ids.push(package_id);
     }
 
-    return transaction("SearchFiles", [ PK_FILTER_INSTALLED, origin_files ],
-                       progress_reporter(0, 5, progress_cb), gather_origin_cb).
-        then(function () {
-            return transaction("RefreshCache", [ true ],
-                               progress_reporter(5, 70, progress_cb)).
-                then (function () {
-                    return transaction("GetUpdates", [ 0 ],
-                                       progress_reporter(75, 5, progress_cb), gather_update_cb).
-                        then(function () {
-                            if (update_ids.length > 0)
-                                return transaction("UpdatePackages", [ 0, update_ids ],
-                                                   progress_reporter(80, 20, progress_cb));
-                        });
+    function search_origin_file_packages() {
+        return transaction("SearchFiles", [ PK_FILTER_INSTALLED, origin_files ],
+                           progress_reporter(0, 5, progress_cb), gather_origin_cb);
+    }
+
+    function refresh_cache() {
+        return transaction("RefreshCache", [ true ],
+                           progress_reporter(5, 70, progress_cb));
+    }
+
+    function maybe_update_origin_file_packages() {
+        return transaction("GetUpdates", [ 0 ],
+                           progress_reporter(75, 5, progress_cb), gather_update_cb).
+            then(function () {
+                if (update_ids.length > 0)
+                    return transaction("UpdatePackages", [ 0, update_ids ],
+                                       progress_reporter(80, 15, progress_cb));
+            });
+    }
+
+    function ensure_collection_packages() {
+        if (collection_packages.length > 0) {
+            return resolve_many("Resolve",
+                                PK_FILTER_ARCH | PK_FILTER_NOT_SOURCE | PK_FILTER_NEWEST | PK_FILTER_NOT_INSTALLED,
+                                collection_packages, progress_reporter(95, 1, progress_cb)).
+                then(function (ids) {
+                    if (ids.length > 0) {
+                        return transaction("InstallPackages", [ 0, ids ], progress_reporter(96, 4, progress_cb)).
```

Comment:
Sorry to be a nitpicker here, but I'd expect downloading and installing this relatively big package (~ 15 MB compressed?) to take much longer than just 4% of the total time. It heeds the usual feeling of "progress bar becomes slower and slower towards the end" :-)

Reply:
Yeah. My reasoning was that normally it is already installed... If we give it, say, 20% of the progressbar, the progress almost always jump from 80% to 100% at the end.  We have that problem already with updating this package.

We can check upfront whether it needs to be installed, bu we can't check whether it needs to be updated.  Hmm.